packages/edit-site/src/components/add-new-template/utils.js

```diff
@@ -536,3 +546,64 @@ const useEntitiesInfo = ( entityName, templatePrefixes ) => {
 	);
 	return entitiesInfo;
 };
+
+/**
+ * Helper function to get the Template Hierarchy for a given slug.
+ * We need to Handle special cases here like `front-page`, `singular` and `archive` templates.
+ *
+ * Noting that we always add `index` as the last fallback template.
+ *
+ * @param {string}  templateSlug        The template slug to be created.
+ * @param {Object}  meta                Metadata about the template to be created.
+ * @param {boolean} meta.isCustom       Indicates if a template is custom or part of the template hierarchy.
+ * @param {string}  meta.templatePrefix The template prefix for the created template. This is used to extract the main template type ex. in `taxonomy-books` we extract the `taxonomy`.
+ *
+ * @return {string[]} The template hierarchy.
+ */
+export const getTemplateHierarchy = ( templateSlug, meta ) => {
+	if ( templateSlug === 'index' ) {
+		return [ 'index' ];
+	}
+	if ( meta?.isCustom ) {
+		return [ 'page', 'singular', 'index' ];
+	}
+	if ( templateSlug === 'front-page' ) {
+		return [ 'front-page', 'home', 'index' ];
+	}
+	const templateHierarchy = [ templateSlug ];
+	// Most default templates don't have `templatePrefix` property.
+	if ( meta?.templatePrefix ) {
+		const [ type ] = meta.templatePrefix.split( '-' );
+		// We need these checks because we always add the `templateSlug` above.
+		if ( ! [ templateSlug, type ].includes( meta.templatePrefix ) ) {
+			templateHierarchy.push( meta.templatePrefix );
+		}
+		if ( templateSlug !== type ) {
+			templateHierarchy.push( type );
+		}
+	}
+	// Handle `archive` template.
+	if (
+		templateSlug.startsWith( 'author' ) ||
+		templateSlug.startsWith( 'taxonomy' ) ||
+		templateSlug.startsWith( 'category' ) ||
+		templateSlug.startsWith( 'tag' ) ||
```

Comment:
It's not a big deal because we have already ruled out custom templates at this point, but the following would be stricter:

```suggestion
		templateSlug.startsWith( 'author-' ) ||
		templateSlug.startsWith( 'taxonomy-' ) ||
		templateSlug.startsWith( 'category-' ) ||
		templateSlug.startsWith( 'tag-' ) ||
```

Reply:
We can't do this because we need to handle the templates that have no dash(author.html, tag.html, etc...). So we also need to add `archive` for them.

Reply:
Ah, you're right, dumb me. :)